### PR TITLE
Implementation Plan: Single-Skill Mode for Headless Sessions (#725)

### LINKED ISSUE #725: Single-skill mode: only inject target skill + deps into headless sessions
## Summary

When `run_skill` fires a headless session, `init_session()` currently writes **every** `BUNDLED_EXTENDED` skill (~80+) into the ephemeral session directory. The headless Claude session sees all of them in its system prompt, consuming an estimated **160–320K tokens** of context on skills it will never use. Since each `run_skill` step invokes exactly one skill, we can limit the session to only the target skill plus its declared transitive `activate_deps`.

## Motivation

- **Context savings**: ~95% reduction in skill-related context for standalone skills (from ~80 skills down to 1). Even multi-dep skills like `make-plan` (needs ~15 skills via the `arch-lens` pack) see major savings.
- **Reduced confusion**: The model won't be distracted by dozens of irrelevant skill descriptions.
- **Faster session setup**: Fewer files to write to the ephemeral tmpfs directory.

### Core change: `allow_only` parameter on `init_session()`

Add `allow_only: frozenset[str] | None = None` to both:
- `DefaultSessionSkillManager.init_session()` in `session_skills.py`
- `SessionSkillManager` Protocol in `_type_protocols.py`

When set, the skill-write loop (line 326) adds a pre-check before `_should_inject_skill()`:
```python
if allow_only is not None and skill_info.name not in allow_only:
    continue
```

When `None` (default), existing behavior is preserved — cook sessions and any other callers are unaffected.

### Pre-compute the allow-set in `run_skill`

In `tools_execution.py`, after `resolve_target_skill()` returns `target_name`:

1. Read the target skill's `activate_deps` from frontmatter
2. Expand pack refs (e.g. `[arch-lens]` → all 13 arch-lens skill names) using skill categories from the provider
3. Recurse into each dep's own `activate_deps` (e.g. each arch-lens declares `[mermaid]`)
4. Collect: `allow_only = {target_name} ∪ transitive_closure`
5. Pass `allow_only` to `init_session()`

### New utility: pack name expansion

No utility currently exists to expand a pack name (e.g. `"arch-lens"`) to skill names without an ephemeral dir on disk. Add a helper like:
```python
def expand_deps_to_skill_names(
    deps: list[str], provider: SkillsDirectoryProvider
) -> frozenset[str]:
    """Expand activate_deps entries (pack names + skill names) to skill name set."""
    ...
```

This needs to handle the recursive case (deps of deps) and be cycle-safe.

### Priority system preserved

The existing priority system is unchanged:
1. **Project-local skills** (highest) — `_should_inject_skill()` still skips bundled skills that have project-local overrides in `.claude/skills/` or `.autoskillit/skills/`. These arrive via `--add-dir <cwd>` regardless of `allow_only`.
2. **BUNDLED_EXTENDED** — filtered by `allow_only`, written to ephemeral dir, served via `--add-dir`.
3. **BUNDLED** (tier 1) — served via `--plugin-dir`, never written to ephemeral dir (unconditionally skipped as before).

### Cook sessions unaffected

`_cook.py` calls `init_session(cook_session=True)` and never passes `allow_only`. Default `None` preserves full skill injection for interactive sessions.

## Pre-requisite: fix undeclared `activate_deps`

Several skills invoke other skills via the Skill tool but don't declare `activate_deps`. These must be fixed first or alongside, otherwise single-skill mode silently breaks them:

| Skill | Missing `activate_deps` | Impact |
|---|---|---|
| `rectify` | `[arch-lens]` | Silent diagram loss (has partial fallback) |
| `open-integration-pr` | `[arch-lens]` | Silent diagram loss (has graceful degradation) |
| `elaborate-phase` | `[dry-walkthrough]` | **Hard failure** — mandatory validation step skipped |
| `make-arch-diag` | `[arch-lens]` | **Hard failure** — entire skill purpose is arch-lens + mermaid |

## Validation rule: enforce `activate_deps` completeness

Add a validation rule (test or recipe rule) that scans all SKILL.md files and enforces: **if a skill body contains a Skill tool invocation pattern targeting `/autoskillit:<name>`, then `<name>` (or its containing pack) must appear in the skill's `activate_deps` transitive closure.** This prevents future regressions.

Scope: only Skill tool invocations matter. References to skills via `run_skill` MCP tool are irrelevant because `run_skill` is hard-blocked inside headless sessions (`_require_not_headless`), so those calls already fail regardless of which skills are in the session directory.

## Test plan

- [ ] `allow_only=frozenset({"skill-a"})` — only that skill written to ephemeral dir
- [ ] `allow_only=frozenset()` — no skills written (edge case)
- [ ] `allow_only=None` — existing behavior preserved (regression guard)
- [ ] Interaction: skill in `allow_only` AND in `effective_disabled` — decide and test precedence (recommend: disabled wins, since explicit disables are intentional)
- [ ] Interaction: skill in `allow_only` but project-local override exists — channel deduplication still skips it
- [ ] Pack expansion: `activate_deps: [arch-lens]` correctly expands to all 13 arch-lens skills + mermaid transitively
- [ ] Transitive deps: skill A → dep B → dep C all present in allow-set
- [ ] Cycle safety: circular `activate_deps` don't cause infinite recursion in pre-computation
- [ ] Cook session: `allow_only` defaults to `None`, full injection preserved
- [ ] `activate_skill_deps()` still works correctly on the reduced skill set
- [ ] Undeclared-dep validation rule catches all 4 known missing declarations
- [ ] End-to-end: `run_skill /autoskillit:investigate` creates session with only `investigate` skill

## Estimated skill categorization

**~45 standalone skills** (single-skill sessions, `allow_only = {self}`):
All audit-*, investigate, rectify (after fix), review-approach, dry-walkthrough, implement-worktree-no-merge, retry-worktree, resolve-failures, resolve-merge-conflicts, review-pr, resolve-review, prepare-pr, compose-pr, triage-issues, analyze-prs, merge-pr, mermaid, diagnose-ci, report-bug, and others.

**~8 multi-dep skills** (small allow-sets):
- `make-plan` → `{make-plan} ∪ arch-lens(13) ∪ {mermaid}` = ~15 skills
- `rectify` → same as make-plan (after fix)
- `open-integration-pr` → same (after fix)
- `elaborate-phase` → `{elaborate-phase, dry-walkthrough}` = 2 skills
- `make-arch-diag` → `{make-arch-diag} ∪ arch-lens(13) ∪ {mermaid}` = ~15 skills
- `make-experiment-diag` → `{make-experiment-diag} ∪ exp-lens(18) ∪ {mermaid}` = ~20 skills

## Labels

`recipe:implementation`

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

`run_skill` currently writes every BUNDLED_EXTENDED skill (~80) into the headless
session's ephemeral skill directory, costing ~160–320K tokens per session in
system-prompt overhead. Each `run_skill` step invokes exactly one skill, so the
session only needs that target plus its declared `activate_deps` transitive closure.

This plan adds:

1. An `allow_only: frozenset[str] | None = None` parameter on
   `DefaultSessionSkillManager.init_session()` and the `SessionSkillManager` Protocol
   that filters the skill-write loop. `None` preserves existing behavior (cook
   sessions and other callers are unaffected).
2. A `compute_skill_closure(target_name)` method on the manager (and a top-level
   helper in `session_skills.py`) that walks `activate_deps` recursively, expanding
   PACK_REGISTRY entries via `categories` frontmatter and recursing into bare skill
   names. Cycle-safe.
3. Wiring in `tools_execution.run_skill` so that when `resolve_target_skill` returns
   a `target_name`, the closure is computed and passed as `allow_only` into
   `init_session`. When the prompt is not a slash command, `allow_only` stays `None`.
4. Frontmatter fixes for four skills with undeclared Skill-tool invocations:
   `rectify`, `open-integration-pr`, `elaborate-phase`, `make-arch-diag`.
5. A contracts test that scans every bundled SKILL.md body for Skill-tool
   invocation patterns and asserts the invoked skill name is in the SKILL.md's
   declared `activate_deps` transitive closure.

### Process Flow Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 40, 'rankSpacing': 52, 'curve': 'basis'}}}%%
flowchart TB
    %% CLASS DEFINITIONS %%
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;

    %% TERMINALS %%
    START([START<br/>━━━━━━<br/>MCP tool call:<br/>run_skill])
    COMPLETE([COMPLETE<br/>━━━━━━<br/>SkillResult → JSON])
    ERROR([ERROR<br/>━━━━━━<br/>gate_error JSON])

    subgraph Guards ["● Guard Chain — tools_execution.py"]
        direction TB
        G1["_require_not_headless()<br/>━━━━━━━━━━<br/>Reject nested headless calls<br/>(AUTOSKILLIT_HEADLESS=1)"]
        G2["_require_enabled()<br/>━━━━━━━━━━<br/>Global gate policy check"]
        G3["_validate_skill_command()<br/>━━━━━━━━━━<br/>Must start with '/' slash<br/>No prose commands"]
        G4{"cwd absolute?<br/>━━━━━━━━━━<br/>Reject relative paths"}
    end

    subgraph Resolution ["● Skill Resolution — tools_execution.py"]
        direction TB
        R1["resolve_target_skill()<br/>━━━━━━━━━━<br/>skill_resolver resolves<br/>slash command → target_name"]
        R2{"target_name<br/>present?<br/>━━━━━━━━━━<br/>resolver returned name"}
    end

    subgraph Closure ["● Closure Computation — session_skills.py"]
        direction TB
        C1["compute_skill_closure(target_name)<br/>━━━━━━━━━━<br/>BFS init: queue=[target_name]<br/>visited=set(), resolved=set()"]
        C2{"dep type?<br/>━━━━━━━━━━<br/>pack key vs bare name"}
        C3["Expand pack via<br/>_build_pack_index()<br/>━━━━━━━━━━<br/>All category members<br/>added to queue"]
        C4["Enqueue bare skill<br/>━━━━━━━━━━<br/>if not yet visited<br/>unknown names silently skipped"]
        C5["frozenset allow_only<br/>━━━━━━━━━━<br/>target + all transitive deps"]
    end

    subgraph SessionInit ["● Session Init — session_skills.py / ● SessionSkillManager protocol"]
        direction TB
        SI1["init_session(session_id, allow_only=…)<br/>━━━━━━━━━━<br/>Create ephemeral dir:<br/>/dev/shm/autoskillit-sessions/<session_id>/"]
        SI2{"● Stage 1:<br/>allow_only filter<br/>━━━━━━━━━━<br/>skill in allow_only?"}
        SI3{"Stage 2:<br/>_should_inject_skill()<br/>━━━━━━━━━━<br/>BUNDLED / overridden<br/>/ pack-disabled?"}
        SI4["Write SKILL.md<br/>━━━━━━━━━━<br/>tier2 skills get<br/>disable-model-invocation: true<br/>injected into frontmatter"]
        SI5["ValidatedAddDir<br/>━━━━━━━━━━<br/>Ephemeral session root<br/>→ skill_add_dirs"]
    end

    subgraph DepActivation ["● Dep Activation — session_skills.py"]
        direction TB
        DA1["activate_skill_deps(session_id, target_name)<br/>━━━━━━━━━━<br/>Input validation (path traversal guard)<br/>→ _activate_with_deps(…, activated={})"]
        DA2["Read SKILL.md from disk<br/>━━━━━━━━━━<br/>_remove_disable_model_invocation()<br/>atomic_write if changed"]
        DA3{"dep type?<br/>━━━━━━━━━━<br/>pack vs bare<br/>(activated set guards cycles)"}
        DA4["_activate_pack_deps()<br/>━━━━━━━━━━<br/>Scan on-disk skills dir<br/>match by category → recurse"]
        DA5["Recurse bare dep<br/>━━━━━━━━━━<br/>_activate_with_deps(_is_root=False)<br/>also strips %%ORDER_UP%% from body"]
    end

    subgraph Execution ["● HeadlessExecutor.run() — headless.py / ● protocol"]
        direction TB
        E1["executor.run(resolved_command, cwd,<br/>add_dirs=[session_root, …])<br/>━━━━━━━━━━<br/>Launches headless Claude session<br/>Only closure-filtered skills visible"]
        E2{"SkillResult<br/>outcome?<br/>━━━━━━━━━━<br/>success / failure"}
    end

    ContractTest["★ test_activate_deps_completeness.py<br/>━━━━━━━━━━<br/>Sweeps all SKILL.md files<br/>Asserts every Skill-tool invocation<br/>has matching activate_deps declaration"]

    %% MAIN FLOW %%
    START --> G1
    G1 -->|"AUTOSKILLIT_HEADLESS=1"| ERROR
    G1 -->|"not headless"| G2
    G2 -->|"disabled"| ERROR
    G2 -->|"enabled"| G3
    G3 -->|"no slash / empty"| ERROR
    G3 -->|"valid"| G4
    G4 -->|"relative path"| ERROR
    G4 -->|"absolute"| R1
    R1 --> R2
    R2 -->|"no target_name<br/>(no resolver)"| SI1_null["init_session<br/>allow_only=None<br/>━━━━━━━━━━<br/>All skills injected<br/>(unfiltered mode)"]
    SI1_null --> E1
    R2 -->|"target_name resolved"| C1
    C1 --> C2
    C2 -->|"pack key"| C3
    C2 -->|"bare name"| C4
    C3 --> C2
    C4 --> C2
    C2 -->|"queue empty"| C5
    C5 --> SI1
    SI1 --> SI2
    SI2 -->|"not in allow_only<br/>→ log debug skip"| SI2
    SI2 -->|"in allow_only"| SI3
    SI3 -->|"BUNDLED / override<br/>/ pack-disabled"| SI2
    SI3 -->|"inject"| SI4
    SI4 --> SI2
    SI2 -->|"all skills iterated"| SI5
    SI5 --> DA1
    DA1 --> DA2
    DA2 --> DA3
    DA3 -->|"pack dep"| DA4
    DA3 -->|"bare dep"| DA5
    DA4 --> DA3
    DA5 --> DA3
    DA3 -->|"deps exhausted"| E1
    E1 --> E2
    E2 -->|"success"| COMPLETE
    E2 -->|"failure"| ERROR

    %% CONTRACT TEST RELATIONSHIP %%
    ContractTest -.->|"verifies activate_deps<br/>declarations are complete"| DA1

    %% CLASS ASSIGNMENTS %%
    class START,COMPLETE,ERROR terminal;
    class G1,G2,G3 detector;
    class G4,R2,C2,SI2,SI3,DA3,E2 stateNode;
    class R1 phase;
    class C1,C3,C4,C5 phase;
    class SI1,SI4,SI5 handler;
    class SI1_null handler;
    class DA1,DA2,DA4,DA5 handler;
    class E1 phase;
    class ContractTest newComponent;
```

Closes #725

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/impl-20260411-111723-020927/.autoskillit/temp/make-plan/single_skill_mode_plan_2026-04-11_112212.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | uncached | output | cache_read | cache_write | count | time |
|------|----------|--------|------------|-------------|-------|------|
| plan | 264 | 29.8k | 1.7M | 123.3k | 1 | 7m 25s |
| verify | 240 | 18.6k | 1.6M | 78.3k | 1 | 4m 55s |
| implement | 735 | 56.8k | 7.6M | 172.9k | 1 | 18m 48s |
| prepare_pr | 60 | 4.9k | 166.3k | 33.8k | 1 | 1m 15s |
| run_arch_lenses | 114 | 12.8k | 386.2k | 77.4k | 2 | 6m 4s |
| compose_pr | 67 | 5.7k | 205.5k | 21.8k | 1 | 1m 30s |
| **Total** | 1.5k | 128.6k | 11.6M | 507.5k | | 40m 0s |

### CODE REVIEW

File: src/autoskillit/server/tools_execution.py [hunk 1]
```diff
@@ -259,13 +259,18 @@ async def run_skill(
 
     skill_add_dirs: list[ValidatedAddDir] = []
     if tool_ctx.session_skill_manager is not None:
+        allow_only: frozenset[str] | None = None
+        if target_name:
+            allow_only = tool_ctx.session_skill_manager.compute_skill_closure(target_name)
```
Comment (Trecek) [inline src/autoskillit/server/tools_execution.py]: [critical] bugs: compute_skill_closure returns frozenset() (empty) when target_name does not resolve to a real skill. The result is passed directly to init_session as allow_only with no empty-check. An empty frozenset causes allow_only to be non-None, so the filter at session_skills.py L389 activates and blocks ALL skills from being injected, silently producing a broken session instead of falling back to full skill injection or raising an error. Fix: treat an empty closure as a resolution failure — set allow_only=None when the closure is empty.

File: tests/workspace/test_session_skills.py [hunk 2]
```diff
@@ -804,3 +804,238 @@ def test_activate_deps_preserves_marker_in_root_skill(self, tmp_path: Path) -> N
...
+    def test_closure_standalone_returns_only_self(self, tmp_path: Path) -> None:
+        from autoskillit.workspace.session_skills import compute_skill_closure
+
+        provider = _make_synthetic_provider(tmp_path, {"lone": {}})
+        assert compute_skill_closure("lone", provider) == frozenset({"lone"})
+
+    def test_closure_pack_dep_expands_to_pack_members(self) -> None:
+        from autoskillit.workspace.session_skills import compute_skill_closure
+
+        provider = SkillsDirectoryProvider()
```
Comment (Trecek) [inline tests/workspace/test_session_skills.py:984]: [warning] tests: test_closure_pack_dep_expands_to_pack_members uses a real SkillsDirectoryProvider() against bundled disk state, making it a fragile integration test disguised as a unit test. The assertion `len(arch_members) == 13` will break silently if the pack membership changes. This exact assertion is also duplicated verbatim in tests/server/test_tools_execution.py:L1505, meaning both tests need updating together whenever the arch-lens pack changes.

File: tests/server/test_tools_execution.py [hunk 3]
```diff
@@ -1348,3 +1348,158 @@ async def test_run_skill_markers_are_unique_per_invocation(self, tool_ctx):
...
+    from autoskillit.server.tools_execution import run_skill
+
+    await run_skill("/autoskillit:make-plan refactor", "/tmp")
+
+    closure = captured["allow_only"]
+    assert closure is not None
+    assert "make-plan" in closure
+    assert "mermaid" in closure
+    arch_members = {n for n in closure if n.startswith("arch-lens-")}
+    assert len(arch_members) == 13
```
Comment (Trecek) [inline tests/server/test_tools_execution.py]: [warning] tests: Magic number assertion `assert len(arch_members) == 13` hard-codes the current arch-lens pack size. This duplicates the identical assertion in tests/workspace/test_session_skills.py:L989 and ties both tests to a concrete count that will silently become wrong when a skill is added or removed from the pack. The contract being tested (pack expands to all its members) can be expressed without a literal count.

File: tests/workspace/test_session_skills.py [hunk 4]
```diff
@@ -804,3 +804,238 @@ def test_activate_deps_preserves_marker_in_root_skill(self, tmp_path: Path) -> N
...
+        closure = compute_skill_closure("make-plan", provider)
+        assert "make-plan" in closure
+        assert "mermaid" in closure  # transitive via arch-lens-* deps
+        arch_members = {n for n in closure if n.startswith("arch-lens-")}
+        assert len(arch_members) == 13
```
Comment (Trecek) [inline tests/workspace/test_session_skills.py]: [warning] tests: Magic number assertion `assert len(arch_members) == 13` hard-codes pack size against live bundled data, making this test brittle to pack membership changes. Consider asserting `len(arch_members) >= 1` or deriving the expected count from the provider itself.

File: tests/server/test_tools_execution.py [hunk 5]
```diff
@@ -1348,3 +1348,158 @@ async def test_run_skill_markers_are_unique_per_invocation(self, tool_ctx):
...
+
+    await run_skill("/test skill", "/tmp")
+
+    mock_ssm.init_session.assert_called_once()
+    assert mock_ssm.init_session.call_args.kwargs.get("allow_only") is None
+    mock_ssm.compute_skill_closure.assert_not_called()
+
+
+@pytest.mark.anyio
+async def test_run_skill_make_plan_closure_includes_arch_lens_pack(tool_ctx, monkeypatch) -> None:
```
Comment (Trecek) [inline tests/server/test_tools_execution.py:1440]: [warning] tests: test_run_skill_make_plan_closure_includes_arch_lens_pack constructs a real SkillsDirectoryProvider() and DefaultSessionSkillManager hitting the actual filesystem, then wraps them in a _RecordingManager that short-circuits init_session. This hybrid approach can mask integration bugs between the two layers. The assertions on 'mermaid' and the hard-coded arch-lens count also couple this test to specific bundled skill content.

File: src/autoskillit/core/_type_protocols.py [hunk 6]
```diff
@@ -434,8 +434,11 @@ def init_session(
         config: Any | None = None,
         project_dir: Path | None = None,
         recipe_packs: frozenset[str] | None = None,
+        allow_only: frozenset[str] | None = None,
     ) -> ValidatedAddDir: ...
 
+    def compute_skill_closure(self, target_name: str) -> frozenset[str]: ...
```
Comment (Trecek) [inline src/autoskillit/core/_type_protocols.py]: [warning] cohesion: Protocol method uses parameter name 'target_name' while the adjacent existing method 'activate_skill_deps' uses 'skill_name' for the same concept. Within the same protocol interface, parameter names for equivalent concepts should be consistent. Prefer 'skill_name' to match the established convention.

File: src/autoskillit/workspace/session_skills.py [hunk 7]
```diff
@@ -210,6 +210,60 @@ def _should_inject_skill(
...
+    """Build a pack-name → set of member skill names index from the provider."""
+    index: dict[str, set[str]] = {}
+    for skill in provider.list_skills():
+        for cat in skill.categories:
+            index.setdefault(cat, set()).add(skill.name)
+    return index
+
+
+def compute_skill_closure(
+    target_name: str,
```
Comment (Trecek) [inline src/autoskillit/workspace/session_skills.py]: [warning] cohesion: Module-level function 'compute_skill_closure' uses parameter name 'target_name', but the broader codebase uses 'skill_name' for the same concept (e.g., activate_skill_deps). The naming divergence between the free function and the class/protocol layer creates a subtle asymmetry.

File: src/autoskillit/workspace/session_skills.py [hunk 8]
```diff
@@ -210,6 +210,60 @@ def _should_inject_skill(
...
+    provider: SkillsDirectoryProvider,
+) -> frozenset[str]:
+    """Return the transitive activate_deps closure for a target skill.
```
Comment (Trecek) [inline src/autoskillit/workspace/session_skills.py]: [warning] slop: Over-verbose docstring for compute_skill_closure documents internal implementation details (breadth-first, PACK_REGISTRY, visited set) that belong in code comments, not the public docstring. The docstring is longer than necessary for a public API description.

File: tests/contracts/test_activate_deps_completeness.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@pytest.fixture(scope="module")
+def provider() -> SkillsDirectoryProvider:
+    return SkillsDirectoryProvider()
+
+
+def test_make_arch_diag_declares_arch_lens_dep(provider: SkillsDirectoryProvider) -> None:
+    closure = compute_skill_closure("make-arch-diag", provider)
+    assert "make-arch-diag" in closure
```
Comment (Trecek) [inline tests/contracts/test_activate_deps_completeness.py]: [warning] slop: Tautological assertion: 'assert "make-arch-diag" in closure' checks that a skill is in its own closure. compute_skill_closure is documented to include the target itself; asserting this without a comment explaining the invariant being tested reads as accidental boilerplate.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: changes_requested (1 critical, 8 warnings found)

**Review by Trecek — COMMENTED** [review]
AutoSkillit review found 1 critical and 8 warnings. See inline comments. (Owner PR — REQUEST_CHANGES event not available; verdict: changes_requested)

File: src/autoskillit/server/tools_execution.py (hunk 1 above)
Comment (Trecek) [reply src/autoskillit/server/tools_execution.py]: Agreed — fixed in 93ece0c. When compute_skill_closure returns frozenset() (target not resolved), allow_only is now set to None rather than an empty frozenset, preventing the init_session filter from activating and blocking all skills.

File: tests/workspace/test_session_skills.py (hunk 2 above)
Comment (Trecek) [reply tests/workspace/test_session_skills.py:984]: Agreed — fixed in 6bea4b8. Changed assert len(arch_members) == 13 to assert len(arch_members) >= 1 so the test remains valid when the arch-lens pack membership changes.

File: tests/server/test_tools_execution.py (hunk 3 above)
Comment (Trecek) [reply tests/server/test_tools_execution.py]: Agreed — fixed in 6bea4b8. Changed assert len(arch_members) == 13 to assert len(arch_members) >= 1 in both duplicate locations.

File: tests/workspace/test_session_skills.py (hunk 4 above)
Comment (Trecek) [reply tests/workspace/test_session_skills.py]: Agreed — fixed in 6bea4b8. Changed assert len(arch_members) == 13 to assert len(arch_members) >= 1.

File: tests/server/test_tools_execution.py (hunk 5 above)
Comment (Trecek) [reply tests/server/test_tools_execution.py:1440]: Valid observation — flagged for design decision. The _RecordingManager hybrid approach (real closure computation + short-circuited init_session) is a deliberate test design tradeoff to enable end-to-end closure verification without a full filesystem session setup. Restructuring to a pure mock or pure integration test would require architectural changes beyond this PR scope.

File: src/autoskillit/core/_type_protocols.py (hunk 6 above)
Comment (Trecek) [reply src/autoskillit/core/_type_protocols.py]: Agreed — fixed in d48e3c6. Renamed target_name to skill_name in the protocol definition to match the established skill_name convention used by activate_skill_deps.

File: src/autoskillit/workspace/session_skills.py (hunk 7 above)
Comment (Trecek) [reply src/autoskillit/workspace/session_skills.py]: Agreed — fixed in d48e3c6. Renamed target_name to skill_name in both the free function compute_skill_closure and DefaultSessionSkillManager.compute_skill_closure for consistency with activate_skill_deps.

File: src/autoskillit/workspace/session_skills.py (hunk 8 above)
Comment (Trecek) [reply src/autoskillit/workspace/session_skills.py]: Agreed — fixed in d48e3c6. Simplified the docstring to a public API description, removing the BFS/PACK_REGISTRY/visited implementation details.

File: tests/contracts/test_activate_deps_completeness.py (hunk 9 above)
Comment (Trecek) [reply tests/contracts/test_activate_deps_completeness.py]: Agreed — fixed in 6bea4b8. Added an inline comment documenting the contract: compute_skill_closure always includes the target itself in the returned closure.
